server/scsynth/SC_PortAudio.cpp

```diff
@@ -262,39 +273,168 @@ int SC_PortAudioDriver::PortAudioCallback(const void* input, void* output, unsig
+        // if one device is specified, let's try to open another one on matching api
+        TryMatchDeviceSameAPI(inDevice, outDevice, IOType::Input);
+        TryMatchDeviceSameAPI(outDevice, inDevice, IOType::Output);
+
+        bool apisAreDifferent = *inDevice != paNoDevice && *outDevice != paNoDevice
+            && Pa_GetDeviceInfo(*inDevice)->hostApi != Pa_GetDeviceInfo(*outDevice)->hostApi;
+        if (apisAreDifferent) {
+            fprintf(stdout, "Requested devices %s and %s use different API. ", GetDeviceName(*inDevice).c_str(),
+                    GetDeviceName(*outDevice).c_str());
+            *outDevice = Pa_GetHostApiInfo(Pa_GetDeviceInfo(*inDevice)->hostApi)->defaultOutputDevice;
+            fprintf(stdout, "Setting output device to %s.\n", GetDeviceName(*outDevice).c_str());
+        }
+        // check for matching sampleRate or requested sample rate
+        if (*inDevice != paNoDevice && *outDevice != paNoDevice) {
+            const auto in_parameters =
+                MakePaStreamParameters(*inDevice, Pa_GetDeviceInfo(*inDevice)->maxInputChannels, 0);
+            const auto out_parameters =
+                MakePaStreamParameters(*outDevice, Pa_GetDeviceInfo(*outDevice)->maxOutputChannels, 0);
+            if (sampleRate) {
+                // check if devices can support requested SR
+                PaError err = Pa_IsFormatSupported(&in_parameters, &out_parameters, sampleRate);
+                if (err != paNoError) {
+                    fprintf(stdout, "\nRequested sample rate %f for devices %s and %s is not supported.\n", sampleRate,
+                            GetDeviceName(*inDevice).c_str(), GetDeviceName(*outDevice).c_str());
+                    return err;
+                }
+            } else {
+                // if we don't request SR, check if devices have maching SR
+                uint32 inSR = uint32(Pa_GetDeviceInfo(*inDevice)->defaultSampleRate);
+                uint32 outSR = uint32(Pa_GetDeviceInfo(*outDevice)->defaultSampleRate);
+                if (inSR != outSR) {
+                    // if defaults are different, check if both devices can be opened using the OUTPUT's SR
+                    PaError err = Pa_IsFormatSupported(&in_parameters, &out_parameters,
+                                                       Pa_GetDeviceInfo(*outDevice)->defaultSampleRate);
+                    if (err != paNoError) {
+                        fprintf(stdout,
+                                "\nRequested devices %s and %s use different sample rates. "
+                                "Please set matching sample rates "
+                                "in the Windows Sound Control Panel and try again.\n",
+                                GetDeviceName(*inDevice).c_str(), GetDeviceName(*outDevice).c_str());
+                        return err;
+                    }
+                }
+            }
+        }
+
+        // in case we still don't have a proper device, use the default device
```

Comment:
> Server crash with ASIO:

Thanks for spotting this!

> If the user specifies a correct ASIO device, the server has to choose that device for both in and out, regardless of the default.

I tried correcting it. If only one device is specified, and it's a full duplex ASIO device, it will be used for both input and output. Could you test on your setup if the fix works correctly?

Reply:
or "use default devices" ?

Reply:
either works